pylibs/unittests/test_real.py

```diff
@@ -0,0 +1,345 @@
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+# 3. Neither the name of the copyright holder nor the
+#    names of its contributors may be used to endorse or promote products
+#    derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+import logging
+import os
+import queue
+import random
+import socket
+import struct
+import threading
+import time
+import unittest
+from typing import Dict, List
+
+import grpc
+
+from OTNSTestCase import OTNSTestCase
+from otns.cli import errors, OTNS
+from otns.proto import visualize_grpc_pb2
+from otns.proto import visualize_grpc_pb2_grpc
+
+
+class UDPSignaler(object):
+    """Signaler for UDP messages.
+    """
+    def __init__(self, id: int):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.source_addr = ("localhost", 9000 + id)
+        self.dest_addr = ("localhost", 9000)
+        self.sock.bind(self.source_addr)
+
+    def send_message(self, message: str) -> None:
+        data = message.encode("ascii")
+        event_packet = struct.pack("<QBH", 0, 5, len(data)) + data
+        self.sock.sendto(event_packet, self.dest_addr)
+
+    def close(self) -> None:
+        self.sock.close()
+
+
+class GRPCThread(threading.Thread):
+    """Thread to listen for expected gRPC stream response.
+    """
+    def __init__(self,
+                 stream,
+                 contain: List[str], 
+                 not_contain: List[str],
+                 exception_queue: queue.Queue,
+                 time_limit=2):
+        threading.Thread.__init__(self)
+        self.stream = stream
+        self.contain = contain
+        self.not_contain = not_contain
+        self.exception_queue = exception_queue
+        self.time_limit = time_limit
+
+    def run(self):
+        start_time = time.time()
+        try:
+            for event in self.stream:
```

Comment:
can we exit the loop if there is no event coming?

Reply:
The `run` method should immediately return in that case?

Reply:
Should there be a timeout and fail after timeouted?

Reply:
Isn't that what the `time_limit` setting is doing? Or do you mean a timeout to get the next stream message? I believe it has heartbeat.

Reply:
Indeed there is a heartbeat sent every second if I recall correctly. 
If so, timeout is not necesssary.